tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/impl/MetricsHelper.java

```diff
@@ -0,0 +1,289 @@
+            .withControlDependencies(
+                Collections.singletonList(assertBroadcastable(tf, weights, tf.onesLike(values))));
+    return ctf.math.mul(weights, tf.onesLike(values));
+  }
+
+  // alias for mean
+
+  /**
+   * Calculate the mean of the operand, along all axes and <code>keepDims</code> is <code>false
+   * </code>
+   *
+   * @param tf the TensorFlow Ops
+   * @param x the Operand used to calculate the mean
+   * @param <T> the type of the Operand.
+   * @param <Z> the data type for the result
+   * @return the mean of the operand
+   */
+  public static <T extends TType, Z extends TNumber> Operand<Z> mean(Ops tf, Operand<T> x) {
+    return mean(tf, x, null, false);
+  }
+
+  /**
+   * Calculate the mean of the operand, alongside the specified axis with <code>keepDims</code> is
+   * <code>false</code>
+   *
+   * @param tf the TensorFlow Ops
+   * @param x the Operand used to calculate the mean
+   * @param axes Axes to compute the mean.
+   * @param <T> the type of the Operand.
+   * @param <U> the type of the axes.
+   * @param <Z> the data type for the result
+   * @return the mean of the operand, along the specified axes.
+   */
+  public static <T extends TType, U extends TIntegral, Z extends TNumber> Operand<Z> mean(
+      Ops tf, Operand<T> x, Operand<U> axes) {
+    return mean(tf, x, axes, false);
+  }
+
+  /**
+   * Calculate the mean of the operand, along all axes.
+   *
+   * @param tf the TensorFlow Ops
+   * @param x the Operand used to calculate the mean
+   * @param keepDims Indicates whether to keep the dimensions or not. If <code>keepdims</code> is
+   *     <code>false</code>, the rank of the tensor is reduced by 1 for each entry in <code>axes
+   *     </code>. If <code>keepdims</code> is <code>true</code>, the reduced dimensions are retained
+   *     with length 1.
+   * @param <T> the type of the operand
+   * @param <Z> the data type for the result
+   * @return the mean of elements of <code>x</code>.
+   */
+  public static <T extends TType, Z extends TNumber> Operand<Z> mean(
+      Ops tf, Operand<T> x, boolean keepDims) {
+    return mean(tf, x, null, keepDims);
+  }
+
+  /**
+   * Calculate the mean of the operand, alongside the specified axes.
+   *
+   * @param tf the TensorFlow Ops
+   * @param x the Operand used to calculate the mean
+   * @param axes Axes to compute the mean.
+   * @param keepDims Indicates whether to keep the dimensions or not. If `keepdims` is `false`, the
+   *     * rank of the tensor is reduced by 1 for each entry in `axes`. If `keepdims` is `true`, the
+   *     * reduced dimensions are retained with length 1.
+   * @param <T> the data type of the Operand
+   * @param <U> the data type of the axes
+   * @param <Z> the data type for the result
+   * @return the mean of elements of `x`.
+   */
+  @SuppressWarnings({"unchecked", "rawtypes"})
+  public static <T extends TType, U extends TIntegral, Z extends TNumber> Operand<Z> mean(
```

Comment:
This method has multiple mismatches between compile-time types and runtime types, interwoven with uses of the `Z` type parameter:
* The method signature requires the caller to specify `Z`.
* The method makes a runtime decision on what runtime type to store in `Operand<Z> xf`. There's no guarantee that the runtime type will match `Z`.
* If `x` is a `TBool`, the method `tf.dtypes.cast`s a `TFloat32` tensor to a `TBool` vector (is that supported?) and the returned tensor type is guaranteed to be different from the use of `Z` in `Operand<Z> xf`.

Do I understand correctly that this method is intended to return an `Operand<T>`  unless `T` is `TBool`, in which case it intends to return an `Operand<TFloat32>`? If so, then I see no way to express the method's intent in Java's compile-time type system. I see two possibilities:
* Perhaps implement separate methods for `T extends TNumber` versus `TBool` and deal with the difference at the point of call?
* Or perhaps return an `Operand<? extends TNumber>`?

Reply:
Ok, I added booleanMean methods taking` Operand<TBool>` as input, and returning `Operand<TFloat64>`.
```
<U extends TIntegral> Operand<TFloat64> booleanMean(
          Ops tf, Operand<TBool> x, Operand<U> axes, boolean keepDims)
```
I changed all the regular mean operations to <T extends TNumber> and took out the ambiguous <Z extends TNumber>.
```
<T extends TNumber, U extends TIntegral> Operand<T> mean(
      Ops tf, Operand<T> x, Operand<U> axes, boolean keepDims) 
```

Reply:
👍 done -- I raised one more specific issue on the new code.